app/src/main/java/org/mozilla/fenix/perf/StartupFrameworkStartMeasurement.kt

```diff
@@ -0,0 +1,86 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.perf
+
+import android.os.Process
+import android.os.SystemClock
+import kotlin.math.roundToLong
+import org.mozilla.fenix.GleanMetrics.StartupTimeline as Telemetry
+
+/**
+ * A class to measure the time the Android framework executes before letting us execute our own code
+ * for the first time in Application's initializer: this value is captured in the
+ * [Telemetry.frameworkStart] probe.
+ *
+ * Since we cannot execute code at process start, this measurement does not fit within the Glean
+ * Timespan metric start/stop programming model so it lives in its own class.
+ */
+internal class StartupFrameworkStartMeasurement(
+    private val stat: Stat = Stat(),
+    private val telemetry: Telemetry = Telemetry,
+    private val getElapsedRealtimeNanos: () -> Long = SystemClock::elapsedRealtimeNanos
+) {
+
+    private var isMetricSet = false
+
+    private var applicationInitNanos = -1L
+    private var isApplicationInitCalled = false
+
+    fun onApplicationInit() {
+        // This gets called from multiple processes: don't do anything expensive. See call site for details.
+        //
+        // In the main process, there are multiple Application impl so we ensure it's only set by
+        // the first one.
+        if (!isApplicationInitCalled) {
+            isApplicationInitCalled = true
+            applicationInitNanos = getElapsedRealtimeNanos()
+        }
```

Comment:
Question, the `getElapsedRealtimeNanos` is the time since boot, so that means x from when the power was turned on, right?

Reply:
> the getElapsedRealtimeNanos is the time since boot, so that means x from when the power was turned on, right?

Yes.

> If that's the case, what exactly are we timing here?

We capture this as a timestamp to later compare it with the "process start time". When combining these (`appInitTime - processStartTime`), we determine the duration that the framework takes before letting us execute code. Notably, the major thing in there is that dex file time that appears quite large in profiles but may not be as large in non-debuggable apps. This probe is primarily to determine if this is worth worrying about.

Since `processStartTime` is determined from clock ticks rather than `getElapsedRealtimeNanos`, I made sure they're comparable – see the comments for details.